roofit/roofitcore/src/RooRealVar.cxx

```diff
@@ -384,41 +369,46 @@ void RooRealVar::setBins(Int_t nBins, const char* name) {
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-/// Add given binning under name 'name' with this variable. If name is null
-/// the binning is installed as the default binning
+/// Add given binning under name 'name' with this variable. If name is null,
+/// the binning is installed as the default binning.
 void RooRealVar::setBinning(const RooAbsBinning& binning, const char* name)
 {
+  std::unique_ptr<RooAbsBinning> newBinning( binning.clone() );
+
   // Process insert hooks required for parameterized binnings
-  if (!name) {
-    RooAbsBinning* newBinning = binning.clone() ;
+  if (!name || strlen(name) == 0) {
     if (_binning) {
       _binning->removeHook(*this) ;
     }
     newBinning->insertHook(*this) ;
-    _binning.reset(newBinning);
+    _binning = std::move(newBinning);
   } else {
-
-    RooLinkedList* altBinning = binning.isShareable() ? &(sharedProp()->_altBinning) : &_altNonSharedBinning ;
-
-    RooAbsBinning* newBinning = binning.clone() ;
-
     // Remove any old binning with this name
-    RooAbsBinning* oldBinning = (RooAbsBinning*) altBinning->FindObject(name) ;
-    if (oldBinning) {
-      altBinning->Remove(oldBinning) ;
-      oldBinning->removeHook(*this) ;
-      delete oldBinning ;
+    auto sharedProps = sharedProp();
+    auto item = sharedProps->_altBinning.find(name);
+    if (item != sharedProps->_altBinning.end()) {
+      item->second->removeHook(*this);
+      if (sharedProps->_ownBinnings)
+        delete item->second;
+
+      sharedProps->_altBinning.erase(item);
+    }
+    auto item2 = _altNonSharedBinning.find(name);
```

Comment:
This can be a change in behavior: before, the binning's "sharability" was checked (`binning.isShareable() ? &(sharedProp()->_altBinning) : &_altNonSharedBinning`), while now you remove it potentially from both. Is this intended?

Reply:
Yes, if it's in both, there's a problem. Doesn't happen in practice, I assume.